docs/user-guide/plugins.md

```diff
@@ -1,3 +1,184 @@
+hooks.main.onReady.add((wrap, element) => {
+    console.log("main", "onReady");
+    if (location.search === '?debug') {
+    console.debug("main", "wrap", wrap);
+    console.debug("main", "element", element);
+    }
+});
+hooks.main.onBeforeRun.add(() => {
+    console.log("main", "onBeforeRun");
+});
+hooks.main.codeBeforeRun.add('print("main", "codeBeforeRun")');
+hooks.main.codeAfterRun.add('print("main", "codeAfterRun")');
+hooks.main.onAfterRun.add(() => {
+    console.log("main", "onAfterRun");
+});
+
+// Use the `worker` attribute on hooks do define plugins that run on workers
+hooks.worker.onReady.add((wrap, xworker) => {
+    console.log("worker", "onReady");
+    if (location.search === '?debug') {
+    console.debug("worker", "wrap", wrap);
+    console.debug("worker", "xworker", xworker);
+    }
+});
+
+hooks.worker.onBeforeRun.add(() => {
+    console.log("worker", "onBeforeRun");
+});
+
+hooks.worker.codeBeforeRun.add('print("worker", "codeBeforeRun")');
+hooks.worker.codeAfterRun.add('print("worker", "codeAfterRun")');
+hooks.worker.onAfterRun.add(() => {
+    console.log("worker", "onAfterRun");
+});
+```
+
+That's it.
+
+
+## Plugins API
+
+As mentioned above, PyScript Plugins API exposes a set of hooks that can be used to intercept
+specific events in the lifecycle of a PyScript application and add or modify specific features
+of the platform itself. To better understand how it works it's important to understand the concepts
+around a PyScript application and plugins.
+
+### Code Execution Methods
+
+There are 2 mains PyScript Applications can execute code: the browser main
+[thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) and on
+[web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers).
+
+We highly recommend users to independently search the difference between the 2 methods to fully understand
+the difference and consequences but here's a short summary:
+
+* main thread: executing code in the browser main thread means the code is being executed in the same
+process where the browser processes user events and paints. By default, the browser uses this single thread
+to run all the JavaScript code in your page, as well as to perform layout, reflows, and garbage collection.
+This means that long-running code or blocking calls can or will block the thread, leading to an unresponsive
+page and a bad user experience.
+* web workers: code executed in workers actually run on "background" threads. This means the code can perform
+tasks without interfering with the user interface or other operations being perfomed in the main thread. While
+this adds great flexibility it's important to understand that workers actually have limited capabilities when
+comparing to code executed on the main thread. For instace, while PyScript offers a DOM API that actually can
+be used in web workers the browser, by default, does not allow DOM operation in workers.
```

Comment:
this is misleading ... we offer both `window` and `document` in the `pyscript` module that's available in either the *main* thread or the *worker* one. The non working with *DOM* namespace is the pyodide or MicroPython *js* one, which doesn't have a document at all in workers but the whole point of our `pyscript` module is to make this difference almost non existent.

Reply:
Good point. I'll try to rephrase it better